hazelcast/src/main/java/com/hazelcast/collection/impl/queue/operations/QueueMergeOperation.java

```diff
@@ -21,50 +21,80 @@
+ * Merges a {@link QueueMergeTypes} for split-brain healing with a {@link SplitBrainMergePolicy}.
  *
  * @since 3.10
  */
 public class QueueMergeOperation extends QueueBackupAwareOperation implements MutatingOperation {
 
-    private SplitBrainMergePolicy<Data, QueueMergeTypes> mergePolicy;
-    private List<QueueMergeTypes> mergingValues;
+    private SplitBrainMergePolicy<Collection<Object>, QueueMergeTypes> mergePolicy;
+    private QueueMergeTypes mergingValue;
 
-    private transient Map<Long, Data> valueMap;
+    private transient Collection<QueueItem> backupCollection;
+    private transient boolean shouldBackup;
 
     public QueueMergeOperation() {
     }
 
-    public QueueMergeOperation(String name, SplitBrainMergePolicy<Data, QueueMergeTypes> mergePolicy,
-                               List<QueueMergeTypes> mergingValues) {
+    public QueueMergeOperation(String name, SplitBrainMergePolicy<Collection<Object>, QueueMergeTypes> mergePolicy,
+                               QueueMergeTypes mergingValue) {
         super(name);
         this.mergePolicy = mergePolicy;
-        this.mergingValues = mergingValues;
+        this.mergingValue = mergingValue;
     }
 
     @Override
     public void run() {
-        QueueContainer queueContainer = getContainer();
-        valueMap = createHashMap(mergingValues.size());
-        for (QueueMergeTypes mergingValue : mergingValues) {
-            QueueItem mergedItem = queueContainer.merge(mergingValue, mergePolicy);
-            if (mergedItem != null) {
-                valueMap.put(mergedItem.getItemId(), mergedItem.getData());
+        QueueContainer container = getContainer();
+        boolean currentCollectionIsEmpty = container.getItemQueue().isEmpty();
+        long currentItemId = container.getCurrentId();
+
+        backupCollection = merge(container, mergingValue, mergePolicy);
+        shouldBackup = currentCollectionIsEmpty != backupCollection.isEmpty() || currentItemId != container.getCurrentId();
+    }
+
+    private Queue<QueueItem> merge(QueueContainer container, QueueMergeTypes mergingValue,
+                                   SplitBrainMergePolicy<Collection<Object>, QueueMergeTypes> mergePolicy) {
+        SerializationService serializationService = getNodeEngine().getSerializationService();
+        serializationService.getManagedContext().initialize(mergingValue);
+        serializationService.getManagedContext().initialize(mergePolicy);
+
+        Queue<QueueItem> existingItems = container.getItemQueue();
+        QueueMergeTypes existingValue = createMergingValue(serializationService, existingItems);
+        Collection<Object> newValues = mergePolicy.merge(mergingValue, existingValue);
+
+        if (isEmpty(newValues)) {
+            if (existingValue != null) {
+                container.clear();
             }
+            getQueueService().destroyDistributedObject(name);
```

Comment:
what was the reason for destroy?

Reply:
I see all items are sent by merging side. How scalable is this approach if i have a long queue merging? Seems easily can get OOME?

Reply:
We concluded that this is not much different than when e.g. a backup op is missed and the anti entropy sends the entire collection to the backup replica, which then contains in memory at one point in time :
- the local collection
- the collection sent from the partition owner
- the serialised collection sent from the partition owner in the input stream